storage/include/sparse_merkle/base_types.h

```diff
@@ -0,0 +1,129 @@
+// Concord
+//
+// Copyright (c) 2019 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the sub-component's license, as noted in the
+// LICENSE file.
+
+#include <cstdint>
+#include <vector>
+
+#include "assertUtils.hpp"
+
+namespace concord {
+namespace storage {
+namespace sparse_merkle {
+
+// A nibble is 4 bits, stored in the lower bits of a byte.
+class Nibble {
+ public:
+  Nibble(uint8_t byte) {
+    Assert((byte & 0xF0) == 0);
+    data_ = byte;
+  }
+
+  bool get_bit(const size_t bit) const {
+    Assert(bit < 4);
+    return (data_ >> bit) & 1;
+  }
+
+  bool operator==(const Nibble& other) const { return data_ == other.data_; }
+
+  // Promote the underlying uint8_t to a size_t so that it can be used as an
+  // index into arrays and vectors.
+  size_t to_index() const { return data_; }
+
+  // Return the underlying representation
+  uint8_t data() const { return data_; }
+
+ private:
+  // Only the lower 4 bits are used
+  uint8_t data_;
+};
+
+// A Hash is a wrapper around a byte buffer containing a hash
+class Hash {
+ public:
+  // Change the following 2 constants if using a different hash algorithm
+  static const size_t SIZE_IN_BITS = 256;
+  const char* HASH_ALGORITHM = "SHA3-256";
+
+  static const size_t SIZE_IN_BYTES = SIZE_IN_BITS / 8;
+  static const size_t MAX_NIBBLES = SIZE_IN_BITS / 16;
+
+  // This is the hash of the empty string ''. It's used as a placeholder value
+  // for merkle tree hash calculations.
+  //
+  // Change this if the hash algorithm changes
+  static constexpr std::array<uint8_t, SIZE_IN_BYTES> EMPTY = {
+      0xa7, 0xff, 0xc6, 0xf8, 0xbf, 0x1e, 0xd7, 0x66, 0x51, 0xc1, 0x47, 0x56, 0xa0, 0x61, 0xd6, 0x62,
+      0xf5, 0x80, 0xff, 0x4d, 0xe4, 0x3b, 0x49, 0xfa, 0x82, 0xd8, 0x0a, 0x4b, 0x80, 0xf8, 0x43, 0x4a};
```

Comment:
does it make sense to do a static_assert() here, ensuring this is exactly _h('')_?

Reply:
Part of the problem is I don't have the hash function available to do this yet :) I double checked it manually 3 times though, with 2 different implementations of sha3-256.